File: Napakalaki/src/NapakalakiGame/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author carlos
 */
public class SpecificBadConsequence extends BadConsequence{
    
    private ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList();
    private ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();
    
    public SpecificBadConsequence(String text, int levels, ArrayList<TreasureKind> tVisible, 
            ArrayList<TreasureKind> tHidden){
        super(text,levels);
        this.specificVisibleTreasures=tVisible;
        this.specificHiddenTreasures=tHidden;
        
    }
    
    // Metodo que comprueba si una BadConsequence esta vacia
    @Override
    public boolean isEmpty(){
        boolean vacio=false;
        if((specificHiddenTreasures.isEmpty()) && (specificVisibleTreasures.isEmpty()))
            vacio=true;
        
        return vacio;
    }
    
    // Metodo que substrae un tesoro de los tesoros visibles de BadConsequence
    @Override
    public void substractVisibleTreasure(Treasure t){
        specificVisibleTreasures.remove(t);
    }
    
    // Metodo que substrae un tesoro de los tesoros visibles de BadConsequence
    @Override
    public void substractHiddenTreasure(Treasure t){
        specificHiddenTreasures.remove(t);
    }
    
    // Metodo que ajusta
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v,ArrayList<Treasure> h){
        BadConsequence pendingBad;
                
            ArrayList<TreasureKind> arr_nV= new ArrayList();
            ArrayList<TreasureKind> arr_nH= new ArrayList();
               
            for(Treasure t: v){
                arr_nV.add(t.getType());   
            }
                
            for(Treasure t: h){
                arr_nH.add(t.getType());
            }
                
                arr_nV.retainAll(specificVisibleTreasures);
                    
                arr_nH.retainAll(specificHiddenTreasures);
                
                
                pendingBad=new SpecificBadConsequence(this.text, 0, arr_nV, arr_nH);
    
        return pendingBad;
    }
}

File: Napakalaki-Ruby/lib/Specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "Bad_consequence"

module NapakalakiGame
class SpecificBadConsequence < BadConsequence
  
  attr_reader:someVisibleTreasures,:someHiddenTreasures
  
  def initialize(aText, someLevels, someSpecificVisibleTreasures, someSpecificHiddenTreasures)
    super(aText,someLevels)
    
    @someVisibleTreasures=someSpecificVisibleTreasures #Tesoro especifico visible
    @someHiddenTreasures=someSpecificHiddenTreasures   #Tesoro especifico invisible
  end
  
  def isEmpty
    empty=false
    
    if ((@someVisibleTreasures==0) && (@someHiddenTreasures==0))
      empty=true
    else if((@someVisibleTreasures.empty?) && (@someHiddenTreasures.empty?))
            empty=true
         else
            empty=false
         end
    end
    
    return empty
  end
  
  def substractVisibleTreasure(treasure) 
    
    @someVisibleTreasures.delete(treasure.type)
  end
  
  def substractHiddenTreasure(treasure) 
    
    @someHiddenTreasures.delete(treasure.type)
  end
  
  def adjustToFitTreasureLists(v,h)
      # Arrays para pasar de tesoros a tipos de tesoros
      copy_visibles=Array.new
      copy_hidden=Array.new
      
      # Arrays que se van a añadir a la BadConsequence
      nV=Array.new
      nH=Array.new
      
      # Copiamos los tipos de tesoros de v
      for t in v
        copy_visibles << t.type
      end
      
      # Copiamos los tipos de tesoros de h
      for t in h
        copy_hidden << t.type
      end
      
      # Guardamos los valores comunes
      if @someVisibleTreasures!=0
      for t in copy_visibles
        if @someVisibleTreasures.index(t)!=nil then
          nV << t
          @someVisibleTreasures.delete_at(@someVisibleTreasures.index(t))
        end
      end
      
      end
      
      # Guardamos los valores comunes
      if @someHiddenTreasures!=0
      for t in copy_hidden
        if @someHiddenTreasures.index(t)!=nil then
          nH << t
          @someHiddenTreasures.delete_at(@someHiddenTreasures.index(t))
        end
      end
      
      end
      
      pending=SpecificBadConsequence.new(@text,@levels,nV,nH)
    
    
    return pending
  end
  
end
end
